mpdev_cleanup.1
.TH mpdev_cleanup 1 "Jul 13, 2020" "manual"
.SH NAME
.PP
mpdev_cleanup - mpdev_cleanup stats / sticker mpd database entries
.SH SYNOPSIS
.PP
mpdev_cleanup [\f[I]option\f[]]

.SH DESCRIPTION
.PP
\fBmpdev_cleanup\fR(1) cleans the stats.db or sticker.db sqlite database by
removing entries that do not correspond to a valid path to a music file.

It can operate using either of the two modes below

.EX
1. Check the presence of a file on the disk (uses access(2) system call).
2. Check the presence of a file in mpd database (loads mpd database into memory)
.EE

Checking against the mpd(1) database is faster than using the access(2) call.
However, if the mpd(1) database has not been updated, there could be entries
in the mpd(1) database which do not correspond to any music file on the disk.
You can use the mpc(1) update command to update the mpd(1) database before
using the mpdev_cleanup(1) command.

Examples

.EX
# Use mpd database as reference

$ mpdev_cleanup -c -d /var/lib/mpd/MDrive/data/stats.db
$ mpdev_cleanup -C -d /var/lib/mpd/MDrive/data/sticker.db

# Use the filesystem as reference

$ mpdev_cleanup -c -m /var/lib/mpd/Music -d /var/lib/mpd/MDrive/data/stats.db
$ mpdev_cleanup -C -m /var/lib/mpd/Music -d /var/lib/mpd/MDrive/data/sticker.db
.EE

.SH OPTIONS
.TP 3
.B -v
Set verbose output. You can increase verbosity by specifying it multiple times
.RS
.RE

.TP 3
\fB-i\fR \fIIP\fR
IP address of mpd host if different from localhost
.RS
.RE

.TP 3
\fB-p\fR \fIport\fR
port on which mpd(1) is listening if different from 6600
.RS
.RE

.TP 3
\fB-s\fR \fIunix_socket\fR
Unix domain socket in case you don't want to use \fIIP\fR and \fIport\fR
.RS
.RE

.TP 3
\fB-d\fR \fIdb_file_path\fR
full path to the sqlite3(1) stats database
.RS
.RE

.TP 3
\fB-m\fR \fImusic_directory\fR
Full path to the \fImusic_directory\fR in the mpd.conf(5) configuration file.
.RS
.RE

.TP 3
.B -c
Clean entries from the stats.db sqlite database
.RS
.RE

.TP 3
.B -C
Clean entries from the sticker.db sqlite database
.RS
.RE

.SH REPORTING BUGS
.PP
If you find a bug, please report it at https://github.com/mbhangui/mpdev/issues

.SH COPYRIGHT
.PP
Copyright (c) 2020 Manvendra Bhangui <mpdev@indimail.org>
.PD 0
.P
.PD
Free
use of this software is granted under the terms of the GNU General
Public License (GPLv2).
.SH AUTHOR
Manvendra Bhangui <manvendra@indimail.org>

.SH SEE ALSO
mpd(1),
mpc(1),
sqlite3(1),
mpdev(1)
mpdev_update(1),
access(2)


ChangeLog
*
Release 0.2 Start 02/03/2020
02/03/2020
01. Initial Release
02. added update_stats utility to maintain stats database
03. mpdev.c: set SONG_DURATION as seconds since epoch
04. player can now on its own add entries to stat database
14/07/2020
05. mpdev - handle empty playlist with no currentsong playing
06. added mpdev_update, mpdev_cleanup
15/07/2020
07. moved mpdev_update, mpdev_cleanup to libexec/mpdev
08. added mpdev_rename script
16/07/2020
09. mpdev_rename: added --query option
10. mpdev: use read instead of timeoutread
17/07/2020
11. mdpev: added START_TIME, END_TIME variables
12. player: refactored script to update ratings from rompr/sticker/stats
13. player: removed hardcored MySQL connection parameters
14. player: fixed variables getting clobbered by functions
19/07/2020
15. player: display sticker insert/updates after NewSong line
16. mpdev_update: added date_added field
17. mpdev_update: use st_mtime, instead of st_ctime for date_added field
18. mpdplaylist: use date_added field instead of last_modified
20/07/2020
19. mpdev: send status, currentsong command on player event to set
    ELAPSED_TIME
20. added rompr_hist play history utility for rompr
21. renamed rompr_hist to romprhist
21/07/2020
22. player: added get_rompr_playcount_ts() function to get playcount and
    timestamp from Playcounttable
23. player: option to run without sticker, stats enabled
24. player: create temp file for recording original start time
25. player: added error condition for getting rating from Ratingtable.
26. player: BUG. fixed rating tmp filename
27. player: display rating for RompR in logs in RompR scale
22/07/2020
28. added transfer_play
29. player: use AUTO_RATING env variable for automatic rating of unrated songs
30. create_service: fix for raspberrypi, dietpi
24/07/2020
31. mpdev_update: removed requirement of -m option for update mode
32. player: exit end-song if END_TIME is not set
33. player: fixed display playcount number in verbose output
28/07/2020
34. mpdev_cleanup.c: made -m optional
35. mpdev_update.c: added updation of Disc tag
35. player.in: added zztimestamp filed to Ratingtable
02/08/2020
36. set default values for play_count, rating, date_added fields.
37. use %mtime% tag from mpd datatabase for date_added filed instead of using stat(2) call
03/08/2020
38. mpdplaylist: added --host option to specify mpd host
39. mpdplaylist: added --shuffle option to shuffle playlist
05/08/2020
40. mpdplaylist: renamed --save option to --playlist
41. mpdev_rename: use SQL REPLACE instead of sed
06/08/2020
42. mpdplaylist: renamed --dummy option to --query
08/08/2020
42. mpdev.c, mpdev_cleanup.c: fixed usage/error strings
43. tcpopen.c: added missing return on connect() failure
09/08/2020
44. player: use unix domain socket if MYSQL_SOCKET is defined
10/08/2020
45. player: skip http urls
46. mpdev: skip submit song when starting from pause state
47. create_rpm: added comments to indicated cwd
48. player: replaced /tmp/mpdev with MPDEV_TMPDIR
49. player: exit now-playing when in pause or stop state
50. mpdev.c: set PLAYER_STATE environment variable when starting
51. mpdev.c: set ELAPSED_TIME, DURATION only when available
52. mpdev.c: fixed setting of initial_state
53. mpdev.c: fixed logic of play/pause/stop
11/08/2020
54. mpdev_update.c: adjust date_added for localtime
55. transfer_play: use mpc to play and remove playlist
14/08/2020
56. added playpause script
26/08/2020
57. player: added self-update of scripts in $HOME/.mpdev
27/08/2020
58. player: auto rate if $HOME/.mpdev/AUTO_RATING file is present
59. fixed path of stats.db
03/08/2020
60. mpdplaylist: fixed --clear option
05/08/2020
61. mpdplaylist: added --limit option
08/08/2020
62. mpdev_rename: fixed usage string
07/01/2021
03. create_services: fixed cp
04. mpdplaylist.in: added last_played IS NULL to daysnotheard query


mpdev_cleanup.c
/*
 * $Log: mpdev_cleanup.c,v $
 * Revision 1.3  2020-08-08 11:07:46+05:30  Cprogrammer
 * fixed usage, error messages
 *
 * Revision 1.2  2020-07-28 12:40:52+05:30  Cprogrammer
 * made -m optional
 *
 * Revision 1.1  2020-07-19 18:16:35+05:30  Cprogrammer
 * Initial revision
 *
 *
 * mpdev_cleanup - utility to clean entries from stats/sticker sqlite db
 *
 */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif
#ifdef HAVE_SQLITE3
#include <sqlite3.h>
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#include <timeoutread.h>
#include <timeoutwrite.h>
#include <substdio.h>
#include <stralloc.h>
#include <error.h>
#include <getln.h>
#include <strerr.h>
#include <sgetopt.h>
#include <fmt.h>
#include <env.h>
#include <scan.h>
#include <str.h>
#include "replacestr.h"
#include "tcpopen.h"

#ifndef	lint
static char     sccsid[] = "$Id: mpdev_cleanup.c,v 1.3 2020-08-08 11:07:46+05:30 Cprogrammer Exp mbhangui $";
#endif

ssize_t         safewrite(int, char *, int);
ssize_t         saferead(int, char *, int);

substdio        mpdin, mpdout, ssout, sserr;
static stralloc line = {0}, tmp = {0};
int             timeout = 1200, disk_mode = 0, verbose;
char            method = 0;
static sqlite3 *db, *memdb;
char           *usage =
				"usage: mpdev_cleanup [-i IP/Host | -s unix_socket] [-p port]\n"
				"  -h IP   - IP address of MPD host. default 127.0.0.1\n"
				"  -p port - MPD listening port. default 6600\n"
				"  -s unix - domain socket path\n"
				"  -m path - music directory path\n"
				"  -d path - sqlite3 dbfile path\n"
				"  -D      - Use disk mode\n"
				"  -c      - Clean stats db\n"
				"  -C      - Clean sticker db\n"
				"  -v      - verbose output";

void
die_write(char *arg)
{
	if (db)
		sqlite3_close(db);
	substdio_puts(&sserr, "Requested action aborted: write error");
	if (arg) {
		substdio_put(&sserr, ": ", 2);
		substdio_puts(&sserr, arg);
	}
	substdio_put(&sserr, "\n", 1);
	substdio_flush(&sserr);
	_exit(111);
}

void
out(char *s)
{
	if (substdio_puts(&ssout, s) == -1)
		die_write(0);
}

void
flush()
{
	if (substdio_flush(&ssout) == -1)
		die_write("stdout");
}

void
err_out(char *s)
{
	if (substdio_puts(&sserr, s) == -1)
		die_write(0);
}

void
err_flush()
{
	if (substdio_flush(&sserr) == -1)
		die_write("sserr");
}

void
die_read(char *arg)
{
	if (db)
		sqlite3_close(db);
	if (arg) {
		substdio_put(&sserr, "Requested action aborted: ", 26);
		substdio_puts(&sserr, arg);
		substdio_put(&sserr, ": read error\n", 13);
	} else
		substdio_puts(&sserr, "Requested action aborted: read error\n");
	substdio_flush(&sserr);
	_exit(111);
}

void
die_alarm()
{
	if (db)
		sqlite3_close(db);
	substdio_puts(&sserr, "Requested action aborted: timeout\n");
	substdio_flush(&sserr);
	_exit(111);
}

void
die_nomem()
{
	if (db)
		sqlite3_close(db);
	substdio_puts(&sserr, "mpdev: out of memory\n");
	substdio_flush(&sserr);
	_exit(111);
}

ssize_t
saferead(int fd, char *buf, int len)
{
	int             r;

	if ((r = timeoutread(timeout, fd, buf, len)) == -1) {
		if (errno == error_timeout)
			die_alarm();
	}
	if (r < 0)
		die_read(0);
	return r;
}

ssize_t
safewrite(int fd, char *buf, int len)
{
	int             r;

	if ((r = timeoutwrite(timeout, fd, buf, len)) <= 0)
		die_write(0);
	return r;
}

int
database_init(int transaction_mode, sqlite3_stmt **r_res, sqlite3_stmt **w_res)
{
	char           *sql, *err_msg;

	*r_res = *w_res = (sqlite3_stmt *) 0;
	if (sqlite3_open(":memory:", &memdb) != SQLITE_OK)
		strerr_die2x(111, "sqlite3_open: memdb: ", (char *) sqlite3_errmsg(memdb));
	if (transaction_mode) {
		if (sqlite3_exec(memdb, "BEGIN TRANSACTION", NULL, NULL, &err_msg) != SQLITE_OK) {
			sqlite3_close(memdb);
			strerr_die2x(111, "BEGIN TRANSACTION: ", err_msg);
		}
		if (err_msg)
			sqlite3_free(err_msg);
	}
	sql = "CREATE TABLE IF NOT EXISTS temp(\n"
        "id  INTEGER PRIMARY KEY,\n"
        "uri TEXT UNIQUE NOT NULL\n"
		")";
	if (sqlite3_exec(memdb, sql, 0, 0, &err_msg) != SQLITE_OK) {
		sqlite3_close(memdb);
		strerr_die4x(111, "sqlite3_exec: ", sql, ": ", err_msg);
	}
	if (err_msg)
		sqlite3_free(err_msg);
	sql = "INSERT or IGNORE into temp (uri) values (@uri)";
	if (sqlite3_prepare_v2(memdb, sql, -1, w_res, 0) != SQLITE_OK) {
		sqlite3_close(memdb);
		strerr_die4x(111, "sqlite3_prepare_v2: ", sql, ": ", (char *) sqlite3_errmsg(db));
	}
	sql = "SELECT uri from temp where uri=@uri";
	if (sqlite3_prepare_v2(memdb, sql, -1, r_res, 0) != SQLITE_OK) {
		sqlite3_close(memdb);
		strerr_die4x(111, "sqlite3_prepare_v2: ", sql, ": ", (char *) sqlite3_errmsg(db));
	}
	return (0);
}

void
insert_data(sqlite3_stmt *res, char *stmt, unsigned long *processed, unsigned long *failure)
{
	sqlite3_bind_text(res, 1, stmt, -1, SQLITE_STATIC);
	if (sqlite3_step(res) != SQLITE_DONE)
		*failure += 1;
	else
		*processed += 1;
	sqlite3_clear_bindings(res);
	sqlite3_reset(res);
}

void
database_end(int transaction_mode)
{
	char           *sql, *err_msg;

	sql = "CREATE INDEX IF NOT EXISTS uri on temp(uri);";
	if (sqlite3_exec(memdb, sql, 0, 0, &err_msg) != SQLITE_OK) {
		sqlite3_close(memdb);
		strerr_die4x(111, "sqlite3_exiec: ", sql, ": ", err_msg);
	}
	if (err_msg)
		sqlite3_free(err_msg);
	if (transaction_mode) {
		if (sqlite3_exec(memdb, "END TRANSACTION", 0, 0, &err_msg) != SQLITE_OK) {
			sqlite3_close(memdb);
			strerr_die2x(111, "sqlite3_exec: END TRANSACTION: ", err_msg);
		}
		if (err_msg)
			sqlite3_free(err_msg);
	}
	return;
}

sqlite3_stmt   *
dump_mpd_into_mem(int sock)
{
	int             match;
	char            mpdinbuf[1024], mpdoutbuf[512];
	unsigned long   processed, failure;
	sqlite3_stmt   *r_res, *w_res;

	substdio_fdbuf(&mpdin, saferead, sock, mpdinbuf, sizeof mpdinbuf);
	substdio_fdbuf(&mpdout, safewrite, sock, mpdoutbuf, sizeof mpdoutbuf);
	if (substdio_put(&mpdout, "listallinfo\n", 12) || substdio_flush(&mpdout) == -1)
		die_write("unable to write to mpd");
	database_init(1, &r_res, &w_res);
	for (processed = failure = 0;;) {
		if (getln(&mpdin, &line, &match, '\n') == -1)
			die_read("getln");
		if (!match && line.len == 0)
			break;
		if (!str_diffn(line.s, "OK\n", 3))
			break;
		line.s[--line.len] = 0; /*- remove newline */
		if (!str_diffn(line.s, "file: ", 6))
			insert_data(w_res, line.s + 6, &processed, &failure);
	}
	database_end(1);
	return r_res;
}

static int
callback(void *data, int argc, char **argv, char **column_name)
{
	int             i, j, notfound = 0;
	char           *ptr;

	for(i = 0; i < argc; i++) {
#if 0
		out(column_name[i]);
		out(" ");
		out(argv[i] ? argv[i] : "NULL");
		out("\n");
#endif
		if (i || !argv[i])
			continue;
		switch (disk_mode)
		{
			case 1:
				notfound = access(argv[i], F_OK);
				break;
			default:
				ptr = sqlite3_expanded_sql((sqlite3_stmt *) data);
				tmp.len = 0;
				if (ptr && (!stralloc_copys(&tmp, ptr) || !stralloc_0(&tmp)))
					die_nomem();
				if (sqlite3_bind_text((sqlite3_stmt *) data, 1, argv[i], -1, SQLITE_STATIC) != SQLITE_OK) {
					strerr_warn4("sqlite3_bind_text: ", argv[i], ": ", (char *) sqlite3_errmsg(memdb), 0);
					strerr_warn4("sqlite3_bind_text: ", tmp.len ? tmp.s : "unknown statement", ": ", (char *) sqlite3_errmsg(memdb), 0);
					return 1;
				}
				j = sqlite3_step((sqlite3_stmt *) data);
				notfound = (j == SQLITE_ROW ? 0 : 1);
				sqlite3_clear_bindings((sqlite3_stmt *) data);
				sqlite3_reset((sqlite3_stmt *) data);
				if (j == SQLITE_ERROR) {
					strerr_warn3("sqlite3_step: ", tmp.len ? tmp.s : "unknown statement", ": ", 0);
					return 1;
				}
				break;
		}
		if (notfound) {
			out(method == 1 ? "DELETE from song where uri='" : "DELETE from sticker where uri='");
			if ((j = replacestr(argv[i], "'", "''", &tmp)))
				out(tmp.s);
			else
				out(argv[i]);
			out("';\n");
		}
	}
	return 0;
}

int
main(int argc, char **argv)
{
	int             opt, sock, port_num = 6600;
	char            port[FMT_ULONG], ssoutbuf[512], sserrbuf[512];
	char           *sql, *database = 0, *err_msg, *music_directory = 0,
				   *mpd_socket, *mpd_host, *ptr;
	sqlite3_stmt   *r_res;

	substdio_fdbuf(&ssout, write, 1, ssoutbuf, sizeof(sserrbuf));
	substdio_fdbuf(&sserr, write, 2, sserrbuf, sizeof(sserrbuf));
	if (!(mpd_host = env_get("MPD_HOST")))
		mpd_host = "127.0.0.1";
	mpd_socket = env_get("MPD_SOCKET");
	if (!(ptr = env_get("MPD_PORT")))
		port_num = 6600;
	else
		scan_uint(ptr, (unsigned int *) &port_num);
	database = (char *) 0;
	while ((opt = getopt(argc, argv, "vcCDd:m:h:s:p:")) != opteof) {
		switch (opt) {
		case 'h':
			mpd_host = optarg;
			break;
		case 's':
			mpd_socket = optarg;
			break;
		case 'p':
			scan_uint(optarg, (unsigned int *) &port_num);
			break;
		case 'm':
			music_directory = optarg;
			break;
		case 'd':
			database = optarg;
			break;
		case 'c': /*- clean stats */
			if (!method)
				method = 1;
			else
				method = 3;
			break;
		case 'C': /*- clean sticker */
			if (!method)
				method = 2;
			else
				method = 3;
			break;
		case 'D':
			disk_mode = 1;
			break;
		case 'v':
			verbose++;
			break;
		default:
			strerr_die1x(100, usage);
		}
	}

	if (!database || !method) {
		err_out("You have to specify -d and either -c or -C options\n");
		err_flush();
		strerr_die1x(100, usage);
	}
	if (method == 3) {
		err_out("You cannot specify both -c and -C options\n");
		err_flush();
		strerr_die1x(100, usage);
	}

	if (!disk_mode) {
		if (mpd_socket && port_num != 6600)
			strerr_die1x(100, "you can't specify both socket & port");
		if (mpd_socket && str_diff(mpd_host, "127.0.0.1"))
			strerr_die1x(100, "you can't specify both socket & IP");
		port[fmt_ulong(port, port_num)] = 0;
		if ((sock = tcpopen(mpd_socket ? mpd_socket : mpd_host, 0, port_num)) == -1) {
			if (mpd_socket)
				strerr_die4sys(111, "mpdev_cleanup: tcpopen: ", "socket [", mpd_socket, "]: ");
			else
				strerr_die6sys(111, "mpdev_cleanup: tcpopen: ", "host [", mpd_host, "] port [", port, "]: ");
		}
		r_res = dump_mpd_into_mem(sock);
	} else
		r_res = (sqlite3_stmt *) 0;
	if (sqlite3_open(database, &db) != SQLITE_OK)
		strerr_die3x(111, database, ": ", (char *) sqlite3_errmsg(db));
	if (music_directory && chdir(music_directory))
		strerr_die3sys(111, "unable to chdir to ", music_directory, ": ");
	sql = (method == 1 ? "SELECT uri from song" : "SELECT uri from sticker");
	if (sqlite3_exec(db, sql, callback, r_res, &err_msg) != SQLITE_OK) {
		sqlite3_close(db);
		strerr_die5x(111, database, ": ", sql, ": ", err_msg);
	}
	if (err_msg)
		sqlite3_free(err_msg);
	sqlite3_close(db);
	out("VACUUM;\n");
	out(".quit\n");
	flush();
}
